master/buildbot/status/web/forms.py

```diff
@@ -103,3 +163,50 @@ def content(self, request, cxt):
         else:
             page = ErrorPage(INTERNAL_SERVER_ERROR, "Missing parameters", "Not all parameters were given")
             defer.returnValue(page.render(request))
+
+    def _getSingleArgument(self, args, encoding, name, default = None):
+        if args.has_key(name) and len(args[name]) == 1:
+            return self.decodeFromURL(args[name][0], encoding)
+        return default
+
+    def _overrideForceContextForField(self, defaultProps, scheduler, field, build, buildMaster, builderName):
+
+        if build.properties.hasProperty(field.name):
+            propertyValue = build.properties.getProperty(field.name, None)
+            buildForceContextForSingleFieldWithValue(defaultProps, scheduler, field, buildMaster, builderName, propertyValue)
+
+        if "nested" in field.type:
+            for subfield in field.fields:
+                self._overrideForceContextForField(defaultProps, scheduler, subfield, build, buildMaster, builderName)
+
+    def _overrideCodebaseSubfield(self, defaultProps, scheduler, field, buildMaster, builderName, name, value):
+        subfield = next((x for x in field.fields if x.name == name), None)
+        if not subfield:
+            return
+
+        buildForceContextForSingleFieldWithValue(defaultProps, scheduler, subfield, buildMaster, builderName, value)
+
+    def _overrideForceContextCodebase(self, defaultProps, scheduler, field, build, buildMaster, builderName):
+        source = next((x for x in build.sources if x.codebase == field.codebase), None)
+        if source is None or "nested" not in field.type:
```

Comment:
by moving `nested not in field type` before `source = next`, we can avoid doing that search in `build.sources`